src/azure-cli-core/azure/cli/core/telemetry.py

```diff
@@ -0,0 +1,133 @@
+
+        client.context.application.id = 'Azure CLI'
+        client.context.application.ver = core_version
+        client.context.user.id = hash(getpass.getuser())
+
+        enable(instrumentation_key)
+    except Exception as ex: #pylint: disable=broad-except
+        # Never fail the command because of telemetry, unless debugging
+        if _debugging():
+            raise ex
+
+def user_agrees_to_telemetry():
+    # TODO: agreement, needs to take Y/N from the command line
+    # and needs a "skip" param to not show (for scripts)
+    return True
+
+def log_event(event_name, **kwargs):
+    """
+    IMPORTANT: do not log events with quotes in the name, properties or measurements;
+    those events may fail to upload.  Also, telemetry events must be verified in
+    the backend because successful upload does not guarentee success.
+    """
+    try:
+        event_name = _remove_quotes(event_name)
+        for key, value in kwargs.items():
+            kwargs[key] = _remove_quotes(value)
+
+        source = 'az'
+        if os.environ.get('CONTINUOUS_INTEGRATION') and os.environ.get('TRAVIS'):
+            source = 'CI'
+        elif ARGCOMPLETE_ENV_NAME in os.environ:
+            source = 'completer'
+
+        props = {
+            'time': str(time.time()),
+            'x-ms-client-request-id': APPLICATION.session['headers']['x-ms-client-request-id'],
+            'command': APPLICATION.session.get('command', None),
+            'version': core_version,
+            'source': source
+            }
+
+        if kwargs:
+            props.update(**kwargs)
+
+        data.append({
+            'name': event_name,
+            'properties': props
+            })
+    except Exception as ex: #pylint: disable=broad-except
+        # Never fail the command because of telemetry, unless debugging
+        if _debugging():
+            raise ex
+
+def _remove_quotes(s):
+    if isinstance(s, str):
+        return s.replace("'", '_').replace('"', '_')
+    return s
+
+def _make_safe(s):
+    for c in '$%^&|':
+        s = s.replace(c, '_')
+    return s
+
+def _debugging():
+    return (os.environ.get('CONTINUOUS_INTEGRATION') and os.environ.get('TRAVIS')) \
+        or _DEBUG_TELEMETRY in os.environ
+
+def telemetry_flush():
+    try:
+        client.flush()
+        if not data:
+            return
+
+        subprocess.Popen([sys.executable,
+                          os.path.realpath(__file__),
+                          _make_safe(json.dumps(data))])
+    except Exception as ex: #pylint: disable=broad-except
+        # Never fail the command because of telemetry, unless debugging
+        if _debugging():
+            raise ex
+
+def upload_telemetry(data_to_save):
+    data_to_save = json.loads(data_to_save.replace("'", '"'))
+
+    for record in data_to_save:
+        client.track_event(record['name'],
+                           record['properties'])
+    client.flush()
+    if _debugging():
+        print('telemetry upload complete')
```

Comment:
we can't use the logger here because we aren't running in the az process

---
In reply to: [78662212](https://github.com/Azure/azure-cli/pull/865#discussion_r78662212) [](ancestors = 78662212)